client/genericclient/stream.go

```diff
@@ -0,0 +1,219 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package genericclient
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"runtime"
+
+	"github.com/cloudwego/kitex/client"
+	"github.com/cloudwego/kitex/client/callopt"
+	"github.com/cloudwego/kitex/pkg/generic"
+	"github.com/cloudwego/kitex/pkg/serviceinfo"
+	"github.com/cloudwego/kitex/pkg/streaming"
+)
+
+type ClientStreaming interface {
+	streaming.Stream
+	Send(req interface{}) error
+	CloseAndRecv() (resp interface{}, err error)
+}
+
+type ServerStreaming interface {
+	streaming.Stream
+	Recv() (resp interface{}, err error)
+}
+
+type BidirectionalStreaming interface {
+	streaming.Stream
+	Send(req interface{}) error
+	Recv() (resp interface{}, err error)
+}
+
+func NewStreamingClient(destService string, g generic.Generic, opts ...client.Option) (Client, error) {
+	svcInfo := newClientStreamingServiceInfo(g)
+	return NewStreamingClientWithServiceInfo(destService, g, svcInfo, opts...)
+}
+
+func NewStreamingClientWithServiceInfo(destService string, g generic.Generic, svcInfo *serviceinfo.ServiceInfo, opts ...client.Option) (Client, error) {
+	var options []client.Option
+	options = append(options, client.WithGeneric(g))
+	options = append(options, client.WithDestService(destService))
+	options = append(options, opts...)
+
+	kc, err := client.NewClient(svcInfo, options...)
+	if err != nil {
+		return nil, err
+	}
+	cli := &genericServiceClient{
+		svcInfo: svcInfo,
+		kClient: kc,
+		g:       g,
+	}
+	runtime.SetFinalizer(cli, (*genericServiceClient).Close)
+
+	svcInfo.GenericMethod = func(name string) serviceinfo.MethodInfo {
+		n, err := g.GetMethod(nil, name)
+		var key string
+		if err != nil {
+			// TODO: support fallback solution for binary
+			key = serviceinfo.GenericMethod
+		} else {
+			key = getGenericStreamingMethodInfoKey(n.StreamingMode)
+		}
+		m := svcInfo.Methods[key]
+		return &methodInfo{
+			MethodInfo: m,
+			oneway:     n.Oneway,
+		}
+	}
+
+	return cli, nil
+}
+
+func getGenericStreamingMethodInfoKey(streamingMode serviceinfo.StreamingMode) string {
+	switch streamingMode {
+	case serviceinfo.StreamingClient:
+		return serviceinfo.GenericClientStreamingMethod
+	case serviceinfo.StreamingServer:
+		return serviceinfo.GenericServerStreamingMethod
+	case serviceinfo.StreamingBidirectional:
+		return serviceinfo.GenericBidirectionalStreamingMethod
+	default:
+		return serviceinfo.GenericMethod
+	}
+}
+
+type clientStreamingClient struct {
```

Comment:
Pay attention to the tracer with client streaming which has a special treatment in `client/stream.go`.

Maybe a test to ensure the RPCFinish event is generated after Recv is called.

Reply:
I confirmed that RPCFinish was called after Recv is called with client streaming.